FEAT: Instructor Dashboard - Course Info - API - Add missing content

brianjbuck-wgu opened the pull request:
## Description

This PR updates the Instructor Dashboard API v2 (`CourseInformationSerializer`) with several improvements to support the frontend micro-frontend (MFE) implementation:

  ### Changes Made

  1. **Added `course_run` field** to `CourseInformationSerializer`
  2. **Fixed OpenAPI schema conflict**
     - Added `ref_name` to both `InstructorTaskSerializer` classes (v1 and v2) to prevent schema generation errors (see screenshots below)
  3. **Added `sort_order` field to all tabs**
     - Enables frontend to easily reorder tabs or insert new tabs between existing ones
     - Values: 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110 (increments of 10)
  4. **Added Bulk Email tab**
     - Conditionally displays when `is_bulk_email_feature_enabled()` returns True and user has staff access
     - Sort order: 100 (between Cohorts and Special Exams)
  5. **Fixes Documentation on  GET /api/instructor/v2/courses/{course_id}** 

  ### User Impact

  **User Roles Affected:**
  - **Course Staff, Instructors, and Admins**: Will see the new `course_run` field in API responses and the Bulk Email tab (when enabled)
  - **Developers**: OpenAPI/Swagger documentation now generates correctly without conflicts
---
**OpenAPI schema generation before:**
<img width="2446" height="1047" alt="image" src="https://github.com/user-attachments/assets/a64cd370-8186-42e2-b8b0-16db15e31bda" />

**OpenAPI schema generation after:**
<img width="1077" height="1158" alt="image" src="https://github.com/user-attachments/assets/50494163-99b5-4aa1-9411-02a122fda42c" />
<img width="443" height="316" alt="image" src="https://github.com/user-attachments/assets/0e24e09b-f77c-4b67-8f2f-ab82798c784a" />

---
<img width="1031" height="1277" alt="image" src="https://github.com/user-attachments/assets/5acfa083-05e3-4e86-b2d7-52f6ca26850b" />

brianjbuck-wgu committed:
feat: Adds course_run field to CourseInformationSerializer v2

Adds sort_order field to tabs JSON in the tabs list in CourseInformationSerializer v2 serializer

Move course_run below course_number and also move the serializer method.

Auto-format removing extraneous white space.

Add trailing commas to the instructor course tabs list to improve the diff going forward.

Add comment about sort order

brianjbuck-wgu committed:
fix: Fixes a bug in OpenAPI schema generation

Fixes a bug in OpenAPI schema generation that would fail due to overlapping namespaces between v1 and v2 APIs

brianjbuck-wgu committed:
feat: Adds the bulk_email tab for staff level users in CourseInformationSerializer v2 serializer.

Rename v2 InstructorTaskSerializer to InstructorTaskSerializerV2

Rename v2 CourseInformationSerializer to CourseInformationSerializerV2

openedx-webhooks commented:
Thanks for the pull request, @brianjbuck-wgu!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

brianjbuck-wgu commented:
Fixes: https://github.com/openedx/edx-platform/issues/37713

wgu-jesse-stewart commented on lms/djangoapps/instructor/views/serializers_v2.py:
```diff
@@ -334,6 +360,9 @@ class InstructorTaskSerializer(serializers.Serializer):
     task_input = serializers.CharField()
     task_output = serializers.CharField(allow_null=True)
 
+    class Meta:
```
I think we can remove this after renaming to InstructorTaskSerializerV2

brianjbuck-wgu replied:
Done.

wgu-jesse-stewart commented on lms/djangoapps/instructor/views/serializers_v2.py:
```diff
@@ -320,6 +341,11 @@ def get_analytics_dashboard_message(self, data):
+        return course_id.run if course_id.run is not None else ''
+
 
 class InstructorTaskSerializer(serializers.Serializer):
```
I think we should rename this to InstructorTaskSerializerV2

brianjbuck-wgu replied:
Done.

wgu-jesse-stewart commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -2467,6 +2467,9 @@ class InstructorTaskSerializer(serializers.Serializer):  # pylint: disable=abstr
     duration_sec = serializers.CharField(help_text=_("Task duration information, if known"))
     task_message = serializers.CharField(help_text=_("User-friendly task status information, if available."))
 
+    class Meta:
```
I think we can remove this after renaming to InstructorTaskSerializerV2

brianjbuck-wgu replied:
Done.

wgu-jesse-stewart commented on lms/djangoapps/instructor/views/serializers_v2.py:
```diff
@@ -63,6 +64,7 @@ class CourseInformationSerializer(serializers.Serializer):
     analytics_dashboard_message = serializers.SerializerMethodField(
         help_text="Message about analytics dashboard availability"
     )
+    course_run = serializers.SerializerMethodField(help_text="Course run identifier")
```
I know it doesn't make a big difference, what do you think about moving this value right after course_number?

brianjbuck-wgu replied:
@wgu-jesse-stewart, That would be reasonable if I'm going to be making modifications renaming the serializer anyway.

brianjbuck-wgu replied:
Done.

wgu-jesse-stewart commented on lms/djangoapps/instructor/views/serializers_v2.py:
```diff
@@ -320,6 +341,11 @@ def get_analytics_dashboard_message(self, data):
         """Get analytics dashboard availability message."""
         return get_analytics_dashboard_message(data['course'].id)
 
+    def get_course_run(self, data):
```
same here, what do you think about moving this value right after course_number?

brianjbuck-wgu replied:
Done.

wgu-jesse-stewart reviewed (changes requested):
This is looking really good - I just have a bit of feedback. Nice work

brianjbuck-wgu commented on lms/djangoapps/instructor/views/serializers_v2.py:
```diff
@@ -162,7 +181,8 @@ def get_tabs(self, data):
                 'title': _('Special Exams'),
-                'url': f'{mfe_base_url}/instructor/{str(course_key)}/special_exams'
+                'url': f'{mfe_base_url}/instructor/{str(course_key)}/special_exams',
+                'sort_order': 110
```
@wgu-jesse-stewart, what are your thoughts about adding trailing commas to these to minimize the diff moving forward?

wgu-jesse-stewart replied:
That makes sense to me

brianjbuck-wgu replied:
Done.

brianjbuck-wgu committed:
fix: Fix OpenAPI schema generation

Fix OpenAPI schema generation for GET /api/instructor/v2/courses/{course_id} to properly show documentation.

wgu-jesse-stewart reviewed (approved):
nice work! Thank you for your attention to details on this

feanil commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -2448,7 +2448,7 @@ def post(self, request, course_id):
 
 
-class InstructorTaskSerializer(serializers.Serializer):  # pylint: disable=abstract-method
+class InstructorTaskSerializerV2(serializers.Serializer):  # pylint: disable=abstract-method
```
If the serializer isn' changing, why bother renaming it?

brianjbuck-wgu replied:
@feanil, These serializers are causing naming conflicts in the OpenAPI schema generation. The options are rename the serializers or add `ref_name` to the Meta class with a different name to both the V1 and V2 serializers. The same issue applies to `CourseInformationSerializerV2` you mentioned below.

feanil replied:
Gotcha, ok sounds good to leave renamed in that case.

feanil commented on lms/djangoapps/instructor/views/serializers_v2.py:
```diff
@@ -33,7 +34,7 @@
 
 
-class CourseInformationSerializer(serializers.Serializer):
+class CourseInformationSerializerV2(serializers.Serializer):
```
Since the one change to this serializer is an addative change, I don't think we need to rename this serializer.

feanil commented on lms/djangoapps/instructor/views/serializers_v2.py:
```diff
@@ -88,42 +90,57 @@ def get_tabs(self, data):
                     'title': _('Course Info'),
-                    'url': f'{mfe_base_url}/instructor/{str(course_key)}/course_info'
+                    'url': f'{mfe_base_url}/instructor/{str(course_key)}/course_info',
+                    'sort_order': 10,
```
Can you explain why the sort_order is needed as a field? The tabs list is already an ordered list, why not just rely on the order of the tabs in the list to determine the order that items are displayed?  If the backend has opinions about order in this list, I think we can just insert items at the correct spots in the list.

brianjbuck-wgu replied:
@wgu-jesse-stewart, Can you comment on the sort order?

wgu-jesse-stewart replied:
@feanil Good question. One thing we need to account for is that the tabs list isn’t always a single authoritative ordered array. The Instructor experience can be extended via the Frontend Plugin Framework, where plugins can Insert tab UI into a slot and ordering is handled via an explicit ordering value (priority), not implicit insertion order. 

If we only rely on backend list order, we can’t deterministically interleave backend-defined tabs with plugin-inserted/conditional tabs (and we risk order shifting based on load/config timing). sort_order gives us a stable, cross-source ordering contract so both backend and plugin contributors can target “between X and Y” consistently.

feanil replied:
Awesome, totally makes sense, @brianjbuck-wgu can you put this context in a comment in the backend code where we're adding sort ordering and maybe cross link to the frontend slots that could cause change so that we have this context in the code for future reference.

feanil reviewed (commented):
I think there are some unnecessary changes here but maybe I'm missing something.
